File: 42-TrappingRainWater/42-TrappingRainWater.cpp
// Last updated: 3/30/2026, 2:14:18 PM
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> lefth, righth;
        int maxheight = -1;
        for(int currheight : height){
            if(currheight>=maxheight) maxheight = max(maxheight,currheight);
            lefth.push_back(maxheight);
        }      
        maxheight = -1;
        reverse(height.begin(),height.end());
        for(int currheight : height){
            if(currheight>=maxheight) maxheight = max(maxheight,currheight);
            righth.push_back(maxheight);
        }
        reverse(height.begin(),height.end());
        reverse(righth.begin(),righth.end());
        int watercount = 0;
        int result = 0;
        for(int index = 0 ; index < height.size() ; index++){
            int minheight = min(lefth[index],righth[index]);
            watercount += minheight - height[index];
            
        }
        return watercount;
    }
};

File: 42-TrappingRainWater/42-TrappingRainWater.java
// Last updated: 3/30/2026, 2:14:11 PM
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] lefth = new int[n];
        int[] righth = new int[n];
        int maxheight = 0;
        for(int index = 0 ; index < n ; index++){
            if(height[index] > maxheight) maxheight = Math.max(maxheight,height[index]);
            lefth[index] = maxheight;
        }
        maxheight = 0;
        for(int index = n-1 ; index>=0 ; index--){
            if(height[index] > maxheight) maxheight = Math.max(maxheight,height[index]);
            righth[index] = maxheight;
        }
        int watercount = 0;
        for(int index = 0 ; index < n ; index++){
            watercount += Math.min(lefth[index],righth[index]) - height[index];
        }
        return watercount;
    }
}